cuda_py/cuda/py/_event.py

```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+# SPDX-License-Identifier: LicenseRef-NVIDIA-SOFTWARE-LICENSE
+
+from dataclasses import dataclass
+from typing import Optional
+
+from cuda import cuda
+from cuda.py._utils import check_or_create_options
+from cuda.py._utils import CUDAError
+from cuda.py._utils import handle_return
+
+
+@dataclass
+class EventOptions:
+    disable_timing: Optional[bool] = False
+    busy_waited_sync: Optional[bool] = False
+    support_ipc: Optional[bool] = False
+
+
+class Event:
+
+    __slots__ = ("_handle", "_timing_disabled", "_busy_waited")
+
+    def __init__(self):
+        # minimal requirements for the destructor
+        self._handle = None
+        raise NotImplementedError(
+            "directly creating an Event object can be ambiguous. Please call "
+            "call Stream.record().")
+
+    @staticmethod
+    def _init(options: Optional[EventOptions]=None):
+        self = Event.__new__(Event)
+        # minimal requirements for the destructor
+        self._handle = None
+
+        options = check_or_create_options(EventOptions, options, "Event options")
+        flags = 0x0
+        self._timing_disabled = self._busy_waited = False
+        if options.disable_timing:
+            flags |= cuda.CUevent_flags.CU_EVENT_DISABLE_TIMING
+            self._timing_disabled = True
+        if options.busy_waited_sync:
+            flags |= cuda.CUevent_flags.CU_EVENT_BLOCKING_SYNC
+            self._busy_waited = True
+        if options.support_ipc:
+            raise NotImplementedError("TODO")
+        self._handle = handle_return(cuda.cuEventCreate(flags))
+        return self
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        # Destroy the event.
+        if self._handle:
+            handle_return(cuda.cuEventDestroy(self._handle))
+            self._handle = None
```

Comment:
Q: What's "above"? Do you mean "after `main` exists"?

Reply:
Sorry, before `main` exits.

e.g. in C++:
```
struct Foo {};
int main(void)
{
    // Destructor for foo runs after the C runtime has exited main.
    static Foo foo{};
}
```

Since cuda runtime/driver _also_ set up atexit handlers (that similarly run below main) and across translation units you can't guarantee the order these will be registered (and hence run), if the destruction ends up running after `main` has completed it might be the case that the driver is already shut down, and then `cuEventDestroy` will segfault or similar.

Reply:
Will look into this as part of #141. I feel we may need to dig into CPython internal to find out...